torch/csrc/jit/serialization/export.cpp

```diff
@@ -1218,6 +1218,35 @@ void GraphEncoder::EncodeTypeProto(
   }
 }
 
+static std::string get_little_endian_data(const at::Tensor& t) {
+#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+  return std::string(
+      static_cast<char*>(t.data_ptr()), t.element_size() * t.numel());
+#elif __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
+  const size_t element_size = t.element_size();
+  const size_t num_elements = t.numel();
+
+  std::vector<char> data_copy{
+      static_cast<char*>(t.data_ptr()),
+      static_cast<char*>(t.data_ptr()) + element_size * num_elements};
+
+  for (size_t i = 0; i < num_elements; ++i) {
+    char* start_byte = data_copy.data() + i * element_size;
+    char* end_byte = start_byte + element_size - 1;
+    /* keep swapping */
+    for (size_t count = 0; count < element_size / 2; ++count) {
+      std::swap(*start_byte, *end_byte);
+      ++start_byte;
+      --end_byte;
+    }
+  }
+
+  return std::string(data_copy.data(), element_size * num_elements);
+#else
+#error Unexpected or undefined __BYTE_ORDER__
```

Comment:
How are this var defined across platforms and compiler? Should we assume little endian by default?

Reply:
No, AFAIK we can't assume little endian by default.

Reply:
Can we use this code to robustly detect endianess? https://github.com/onnx/onnx/blob/821d156046a96566e7bb8456696c04716bd27f27/onnx/common/platform_helpers.h#L11-L17

```cpp
inline bool is_processor_little_endian() {
  int num = 1;
  if (*(char*)&num == 1)
    return true;
  return false;
}
```

Reply:
Also:

https://github.com/pytorch/pytorch/blob/653c1564bf4c46d9758000ea70a610ee3699acf1/torch/csrc/api/src/data/datasets/mnist.cpp#L28-L36

and

https://github.com/pytorch/pytorch/blob/653c1564bf4c46d9758000ea70a610ee3699acf1/caffe2/operators/fused_rowwise_nbitfake_conversion_ops.h#L16-L19

Reply:
Yeah, endianness can be detected at runtime. But isn't it better to do it runtime and avoid unnecessary runtime checks?

Reply:
I think all modern compilers define this macro, so I consider it unlikely, but not impossible.